File: src/java/nz/co/searchwellington/controllers/UrlStack.java
package nz.co.searchwellington.controllers;

import javax.servlet.http.HttpServletRequest;

import org.apache.log4j.Logger;

import nz.co.searchwellington.model.SiteInformation;

public class UrlStack {
    
    Logger log = Logger.getLogger(UrlStack.class);
    
    
    private SiteInformation siteInformation;
    
   
    public UrlStack(SiteInformation siteInformation) {
        this.siteInformation = siteInformation;
    }


    public String getExitUrlFromStack(HttpServletRequest request) {
        String url = siteInformation.getUrl() + "/index";
        final String stackUrl = (String) request.getSession().getAttribute("url");        
        if (stackUrl != null) {
            log.debug("Stack url is: " + stackUrl);
            url = (String) siteInformation.getUrl() + stackUrl;
        }
        log.debug("Exit url from stack is: " + url);
        return url;
    }
    
    
    public void setUrlStack(HttpServletRequest request) {        
        String url = request.getPathInfo();
        if (request.getQueryString() != null) {
            url = url + "?" + request.getQueryString();
        }
        setUrlStack(request, url);
    }


	public void setUrlStack(HttpServletRequest request, String url) {
		request.getSession().setAttribute("url", url);
        log.debug("Put url onto the stack: " + url);
	}
    
}

File: src/main/scala/nz/co/searchwellington/controllers/UrlStack.scala
package nz.co.searchwellington.controllers

import jakarta.servlet.http.HttpServletRequest
import nz.co.searchwellington.filters.RequestPath
import nz.co.searchwellington.urls.UrlBuilder
import org.apache.commons.logging.LogFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component class UrlStack @Autowired()(urlBuilder: UrlBuilder) {

  private val log = LogFactory.getLog(classOf[UrlStack])

  def getExitUrlFromStack(request: HttpServletRequest): String = {
    Option(request.getSession.getAttribute("url").asInstanceOf[String]).map { uri =>
      log.debug("Stack url is: " + uri)
      urlBuilder.fullyQualified(uri)
    }.getOrElse{
      urlBuilder.fullyQualified(urlBuilder.getHomeUri)
    }
  }

  def setUrlStack(request: HttpServletRequest): Unit = {
    var url = RequestPath.getPathFrom(request)
    if (request.getQueryString != null) {
      url = url + "?" + request.getQueryString
    }
    setUrlStack(request, url)
  }

  def setUrlStack(request: HttpServletRequest, url: String): Unit = {
    request.getSession.setAttribute("url", url)
    log.debug("Put url onto the stack: " + url)
  }

}